blockchain/upgrade.go

```diff
@@ -1144,6 +1161,97 @@ func initializeGCSFilters(ctx context.Context, db database.DB, genesisHash *chai
 	return nil
 }
 
+// -----------------------------------------------------------------------------
+// The version 3 block index consists of an entry for every known block.  It
+// consists of information such as the block header and information about votes.
+//
+// The serialized key format is:
+//
+//   <block height><block hash>
+//
+//   Field           Type              Size
+//   block height    uint32            4 bytes
+//   block hash      chainhash.Hash    chainhash.HashSize
+//
+// The serialized value format is:
+//
+//   <block header><status><num votes><votes info>
+//
+//   Field              Type                Size
+//   block header       wire.BlockHeader    180 bytes
+//   status             byte                1 byte
+//   num votes          VLQ                 variable
+//   vote info
+//     vote version     VLQ                 variable
+//     vote bits        VLQ                 variable
+//
+// The version 3 block status flags format is:
+//
+//   bit  0   - block payload is stored on disk
+//   bit  1   - block and all of its ancestors have been fully validated
+//   bit  2   - block failed validation
+//   bit  3   - an ancestor of the block failed validation
+//   bits 4-7 - unused
+// -----------------------------------------------------------------------------
+
+// blockIndexEntryV3 represents a version 3 block index database entry.
+type blockIndexEntryV3 struct {
+	header   wire.BlockHeader
+	status   byte
+	voteInfo []blockIndexVoteVersionTuple
+}
+
+// blockIndexEntrySerializeSizeV3 returns the number of bytes it would take to
+// serialize the passed block index entry according to the version 3 format
+// described above.
+func blockIndexEntrySerializeSizeV3(entry *blockIndexEntryV3) int {
+	voteInfoSize := 0
+	for i := range entry.voteInfo {
+		voteInfoSize += serializeSizeVLQ(uint64(entry.voteInfo[i].version)) +
+			serializeSizeVLQ(uint64(entry.voteInfo[i].bits))
+	}
+
+	const blockHdrSize = 180
```

Comment:
Consider a const definition at the top of the file for consistency.

Reply:
I'm a bit torn on this.  I definitely agree that typically constants like this that are used multiple times are best defined at the top of the file, however, this is all upgrade code which means it has to be valid for whatever the values were at that moment in time, and thus unlike the normal case where having a single globally-defined constant is ideal for increasing robustness, having global definitions in upgrade paths leads to more brittle code.

The block header size isn't likely to change anytime soon since it has huge ramifications on mining hardware, but it very well could change at some point and thus it being this exact value for v3 entries is important.

Reply:
You're right, it makes sense to have constants within a local scope for upgrade related code even though there will be some repetition.